### PR TITLE
[zero] refactor low level zero

### PR DESCRIPTION
**Pull request by Gy-Lu**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> If you have any plots/diagrams/screenshots/tables, please attach them here.

Refactor low level zero and improve the compatibility.
Support (shard/no-shard) checkpoint IO.
Support gradient accumulation with no_sync for zero1.
Support passing process group.

### Benchmark

**Bert**

|       | max cuda mem | throughput(sample/s) | params |
| :-----| -----------: | :--------: | :----: |
| ddp | 21.44 GB | 3.0 | 86.8M |
| ddp_fp16 | 16.26 GB | 11.6 | 86.8M |
| gemini | 11.3 GB | 14.6 | 86.8M |
| low_level_zero | 11.29 G | 15.4 | 86.8M |

**AlBert**
|       | max cuda mem | throughput(sample/s) | params |
| :-----| -----------: | :--------: | :----: |
| ddp | OOM |  | |
| ddp_fp16 | OOM |  | |
| gemini | 69.39 G | 1.4 | 218.9M |
| low_level_zero | 56.89 G | 1.4 | 218.9M |

### Convergence Test

Bert-gelu Finetune F1-score=0.88

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

**Comment by ver217** [comment]
Can you becnmark on larger models, like llama-7b and llama-13b and compare to the previous version?

**Comment by Gy-Lu** [comment]
> Can you becnmark on larger models, like llama-7b and llama-13b and compare to the previous version?

OK, I'll do that.
